File: src/com/castlefrog/agl/domains/hex/HexState.java
package com.castlefrog.agl.domains.hex;

import java.io.Serializable;

import com.castlefrog.agl.State;
import sun.management.resources.agent;
import sun.net.www.content.audio.basic;

/*
 * A Hex state consists of a board of hexagon locations
 * that are either empty or have a piece black or
 * white piece.
 */
public final class HexState implements State<HexState>, Serializable {
    private static final int MIN_BOARD_SIZE = 1;

    private byte[][] bitBoards_;
    private byte boardSize_;
    private byte agentTurn_;

    public static final int
        LOCATION_EMPTY = 0,
        LOCATION_BLACK = 1,
        LOCATION_WHITE = 2;

    public HexState(int boardSize,
                    byte[][] bitBoards,
                    int agentTurn) {
        if (boardSize < MIN_BOARD_SIZE) {
            throw new IllegalArgumentException("Invalid board size: " + boardSize);
        }
        boardSize_ = (byte) boardSize;
        bitBoards_ = new byte[bitBoards.length][bitBoards[0].length];
        for (int i = 0; i < bitBoards.length; i += 1) {
            System.arraycopy(bitBoards[i], 0, bitBoards_[i], 0, bitBoards[0].length);
        }
        agentTurn_ = (byte) agentTurn;
    }

    public HexState copy() {
        return new HexState(boardSize_, bitBoards_, agentTurn_);
    }

    public byte[][] getBitBoards() {
        byte[][] bitBoards = new byte[bitBoards_.length][bitBoards_.length];
        for (int i = 0; i < bitBoards_.length; i += 1) {
            System.arraycopy(bitBoards_[i], 0, bitBoards[i], 0, bitBoards_.length);
        }
        return bitBoards;
    }

    public byte[][] getLocations() {
        byte[][] locations = new byte[boardSize_][boardSize_];
        for (int i = 0; i < boardSize_; i += 1) {
            for (int j = 0; j < boardSize_; j += 1) {
                locations[i][j] = (byte) getLocation(i, j);
            }
        }
        return locations;
    }

    public int getLocation(int x, int y) {
        int bitLocation = y * boardSize_ + x;
        int byteLocation = bitLocation / Byte.SIZE;
        if ((bitBoards_[0][byteLocation] & (1 << bitLocation % Byte.SIZE)) != 0) {
            return LOCATION_BLACK;
        } else if ((bitBoards_[1][byteLocation] & ( 1 << bitLocation % Byte.SIZE)) != 0) {
            return  LOCATION_WHITE;
        } else {
            return LOCATION_EMPTY;
        }
    }

    public int getBoardSize() {
        return boardSize_;
    }

    public byte getAgentTurn() {
        return agentTurn_;
    }

    public boolean isLocationEmpty(int x, int y) {
        return getLocation(x, y) == LOCATION_EMPTY;
    }

    public boolean isBoardEmpty() {
        for (int i = 0; i < bitBoards_.length; i += 1) {
            for (int j = 0; j < bitBoards_[0].length; j += 1) {
                if (bitBoards_[i][j] != 0) {
                    return false;
                }
            }
        }
        return true;
    }

    public int getNPieces() {
        int nPieces = 0;
        byte[][] locations = getLocations();
        for (int i = 0; i < locations.length; i += 1) {
            for (int j = 0; j < locations.length; j += 1) {
                if (locations[i][j] != LOCATION_EMPTY) {
                    nPieces += 1;
                }
            }
        }
        return nPieces;
    }

    public void setLocation(int x, int y, int value) {
        int bitLocation = y * boardSize_ + x;
        int byteLocation = bitLocation / Byte.SIZE;
        int byteShift = bitLocation % Byte.SIZE;
        if (value == LOCATION_EMPTY) {
            bitBoards_[0][byteLocation] &= ((1 << byteShift) ^ 0xff);
            bitBoards_[1][byteLocation] &= ((1 << byteShift) ^ 0xff);
        } else if (value == LOCATION_BLACK) {
            bitBoards_[0][byteLocation] |= 1 << byteShift;
            bitBoards_[1][byteLocation] &= ((1 << byteShift) ^ 0xff);
        } else if (value == LOCATION_WHITE) {
            bitBoards_[0][byteLocation] &= ((1 << byteShift) ^ 0xff);
            bitBoards_[1][byteLocation] |= 1 << byteShift;
        }
    }

    public void setAgentTurn(int agentTurn) {
        agentTurn_ = (byte) agentTurn;
    }

    @Override
    public int hashCode() {
        int hashCode = 17;
        hashCode = hashCode * 13 + boardSize_;
        hashCode = hashCode * 23 + agentTurn_;
        for (byte[] row : bitBoards_) {
            for (byte location : row) {
                hashCode = 31 * hashCode + location;
            }
        }
        return hashCode;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof HexState)) {
            return false;
        }
        HexState state = (HexState) object;
        byte[][] bitBoards = state.getBitBoards();
        for (int i = 0; i < bitBoards_.length; i += 1) {
            for (int j = 0; j < bitBoards_.length; j += 1) {
                if (bitBoards_[i][j] != bitBoards[i][j]) {
                    return false;
                }
            }
        }
        return agentTurn_ == state.getAgentTurn() &&
               boardSize_ == state.getBoardSize();
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        for (int i = boardSize_ - 1; i >= 0; i -= 1) {
            for (int j = i; j < boardSize_ - 1; j += 1) {
                output.append(" ");
            }
            for (int j = 0; j < boardSize_; j += 1) {
                if (getLocation(j, i) == LOCATION_BLACK) {
                    output.append("X ");
                } else if (getLocation(j, i) == LOCATION_WHITE) {
                    output.append("O ");
                } else {
                    output.append("- ");
                }
            }
            output.append("\n");
        }
        return output.toString();
    }
}

File: src/main/kotlin/com/castlefrog/agl/domains/hex/HexState.kt
package com.castlefrog.agl.domains.hex

import com.castlefrog.agl.State

data class HexState(
    val boardSize: Int,
    val bitBoards: Array<ByteArray> = Array(2) {
        ByteArray((boardSize * boardSize + java.lang.Byte.SIZE - 1) / java.lang.Byte.SIZE)
    },
    var agentTurn: Byte = TURN_BLACK
) : State<HexState> {

    companion object {
        const val LOCATION_EMPTY = 0
        const val LOCATION_BLACK = 1
        const val LOCATION_WHITE = 2

        const val TURN_BLACK: Byte = 0
        const val TURN_WHITE: Byte = 1

        private const val BYTE_MASK = 0xff
    }

    override fun copy(): HexState {
        val copyBitBoards = Array(bitBoards.size) { bitBoards[it].copyOf() }
        return HexState(boardSize, copyBitBoards, agentTurn)
    }

    val locations: Array<ByteArray>
        get() {
            val locations = Array(boardSize) { ByteArray(boardSize) }
            var i = 0
            while (i < boardSize) {
                var j = 0
                while (j < boardSize) {
                    locations[i][j] = getLocation(i, j).toByte()
                    j += 1
                }
                i += 1
            }
            return locations
        }

    fun getLocation(x: Int, y: Int): Int {
        checkLocationArgs(x, y)
        val index = bitIndex(x, y)
        return when {
            isSet(bitBoards[0], index) -> LOCATION_BLACK
            isSet(bitBoards[1], index) -> LOCATION_WHITE
            else -> LOCATION_EMPTY
        }
    }

    fun isLocationEmpty(x: Int, y: Int): Boolean {
        checkLocationArgs(x, y)
        val index = bitIndex(x, y)
        return !isSet(bitBoards[0], index) && !isSet(bitBoards[1], index)
    }

    private fun isLocationOnBoard(x: Int, y: Int): Boolean {
        return x in 0 until boardSize && y in 0 until boardSize
    }

    val nPieces: Int
        get() = bitBoards[0].indices.sumOf { byteIndex ->
            val occupied = (bitBoards[0][byteIndex].toInt() or bitBoards[1][byteIndex].toInt()) and BYTE_MASK
            Integer.bitCount(occupied)
        }

    fun setLocation(x: Int, y: Int, value: Int) {
        checkLocationArgs(x, y)
        val index = bitIndex(x, y)
        when (value) {
            LOCATION_EMPTY -> {
                clearBit(bitBoards[0], index)
                clearBit(bitBoards[1], index)
            }
            LOCATION_BLACK -> {
                setBit(bitBoards[0], index)
                clearBit(bitBoards[1], index)
            }
            LOCATION_WHITE -> {
                clearBit(bitBoards[0], index)
                setBit(bitBoards[1], index)
            }
        }
    }

    private fun bitIndex(x: Int, y: Int): Int = y * boardSize + x

    private fun isSet(board: ByteArray, bitIndex: Int): Boolean {
        val bits = board[byteIndex(bitIndex)].toInt()
        return (bits and bitMask(bitIndex)) != 0
    }

    private fun setBit(board: ByteArray, bitIndex: Int) {
        val i = byteIndex(bitIndex)
        board[i] = (board[i].toInt() or bitMask(bitIndex)).toByte()
    }

    private fun clearBit(board: ByteArray, bitIndex: Int) {
        val i = byteIndex(bitIndex)
        board[i] = (board[i].toInt() and clearMask(bitIndex)).toByte()
    }

    private fun byteIndex(bitIndex: Int): Int = bitIndex / java.lang.Byte.SIZE

    private fun bitMask(bitIndex: Int): Int = 1 shl (bitIndex % java.lang.Byte.SIZE)

    /** 8-bit mask with the target bit cleared; matches historical `xor 0xff` clearing. */
    private fun clearMask(bitIndex: Int): Int = bitMask(bitIndex) xor BYTE_MASK

    private fun checkLocationArgs(x: Int, y: Int) {
        if (!isLocationOnBoard(x, y))
            throw IllegalArgumentException("(x=$x,y=$y) out of bounds")
    }

    override fun hashCode(): Int {
        var hashCode = 17 + boardSize
        hashCode = hashCode * 19 + bitBoards.contentHashCode()
        hashCode = hashCode * 31 + agentTurn
        return hashCode
    }

    override fun equals(other: Any?): Boolean {
        if (other !is HexState) {
            return false
        }
        if (bitBoards.size != other.bitBoards.size) {
            return false
        }
        for (i in bitBoards.indices) {
            if (bitBoards[i].size != other.bitBoards[i].size) {
                return false
            }
            bitBoards[i].indices
                .filter { other.bitBoards[i][it] != bitBoards[i][it] }
                .forEach { _ -> return false }
        }
        return other.boardSize == boardSize && other.agentTurn == agentTurn
    }

    override fun toString(): String {
        val output = StringBuilder()
        for (i in boardSize - 1 downTo 0) {
            for (j in i..boardSize - 2) {
                output.append(" ")
            }
            for (j in 0 until boardSize) {
                when (getLocation(j, i)) {
                    LOCATION_BLACK -> output.append("X")
                    LOCATION_WHITE -> output.append("O")
                    else -> output.append("-")
                }
                if (j != boardSize - 1) {
                    output.append(" ")
                }
            }
            if (i != 0) {
                output.append("\n")
            }
        }
        return output.toString()
    }
}
